client/relay-chain-minimal-node/src/blockchain_rpc_client.rs

```diff
@@ -0,0 +1,490 @@
+		Ok(result)
+	}
+
+	async fn disputes(
+		&self,
+		at: Hash,
+	) -> Result<
+		Vec<(
+			polkadot_primitives::v2::SessionIndex,
+			polkadot_primitives::v2::CandidateHash,
+			polkadot_primitives::v2::DisputeState<polkadot_primitives::v2::BlockNumber>,
+		)>,
+		ApiError,
+	> {
+		let result = self.rpc_client.parachain_host_staging_get_disputes(at).await?;
+		Ok(result)
+	}
+}
+
+#[async_trait::async_trait]
+impl AuthorityDiscovery<Block> for BlockChainRpcClient {
+	async fn authorities(
+		&self,
+		at: Hash,
+	) -> std::result::Result<Vec<polkadot_primitives::v2::AuthorityDiscoveryId>, sp_api::ApiError> {
+		let result = self.rpc_client.authority_discovery_authorities(at).await?;
+		Ok(result)
+	}
+}
+
+impl BlockChainRpcClient {
+	pub async fn local_listen_addresses(
+		&self,
+	) -> Result<Vec<MultiaddrWithPeerId>, RelayChainError> {
+		let addresses = self.rpc_client.system_local_listen_addresses().await?;
+		tracing::debug!(target: LOG_TARGET, ?addresses, "Fetched listen address from RPC node.");
+
+		let mut result_vec = Vec::new();
+		for address in addresses {
+			match MultiaddrWithPeerId::from_str(&address) {
+				Ok(addr) => result_vec.push(addr),
+				Err(err) => return Err(RelayChainError::GenericError(err.to_string())),
+			}
+		}
+
+		Ok(result_vec)
+	}
+
+	pub async fn import_notification_stream_async(
+		&self,
+	) -> Pin<Box<dyn Stream<Item = Header> + Send>> {
+		self.rpc_client
+			.get_imported_heads_stream()
+			.await
+			.expect("subscribe_all_heads")
+			.boxed()
+	}
+
+	pub async fn finality_notification_stream_async(
+		&self,
+	) -> Pin<Box<dyn Stream<Item = Header> + Send>> {
+		self.rpc_client
+			.get_finalized_heads_stream()
+			.await
+			.expect("imported_headers_stream")
+			.boxed()
+	}
+}
+
+fn block_local<T>(fut: impl Future<Output = T>) -> T {
+	let tokio_handle = tokio::runtime::Handle::current();
+	tokio::task::block_in_place(|| tokio_handle.block_on(fut))
+}
+
+impl HeaderBackend<Block> for BlockChainRpcClient {
+	fn header(
+		&self,
+		id: BlockId,
+	) -> sp_blockchain::Result<Option<<Block as polkadot_service::BlockT>::Header>> {
+		let fetch_header = |hash| block_local(self.rpc_client.chain_get_header(Some(hash)));
+
+		match id {
+			BlockId::Hash(hash) => Ok(fetch_header(hash)?),
+			BlockId::Number(number) => {
+				if let Some(hash) = HeaderBackend::<Block>::hash(self, number)? {
+					Ok(fetch_header(hash)?)
+				} else {
+					Ok(None)
+				}
+			},
+		}
+	}
+
+	fn info(&self) -> Info<Block> {
+		let best_header = block_local(self.rpc_client.chain_get_header(None))
+			.expect("get_header")
+			.unwrap();
+		let genesis_hash = block_local(self.rpc_client.chain_get_head(Some(0))).expect("get_head");
+		let finalized_head =
+			block_local(self.rpc_client.chain_get_finalized_head()).expect("get_head");
+		let finalized_header = block_local(self.rpc_client.chain_get_header(Some(finalized_head)))
+			.expect("get_head")
+			.unwrap();
```

Comment:
No unwraps and better expect messages please.

Reply:
I removed some of the expect messages here, some are left however since `HeaderBackend` methods are not expected to fail currently and we can not handle request errors anyway in this version.